source/extensions/filters/http/composite/filter.cc

```diff
@@ -0,0 +1,188 @@
+}
+} // namespace
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
+  decoded_headers_ = true;
+
+  return delegateFilterAction(delegated_filter_, &StreamDecoderFilter::decodeHeaders,
+                              Http::FilterHeadersStatus::Continue, headers, end_stream);
+}
+
+Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_stream) {
+  return delegateFilterAction(delegated_filter_, &StreamDecoderFilter::decodeData,
+                              Http::FilterDataStatus::Continue, data, end_stream);
+}
+
+Http::FilterTrailersStatus Filter::decodeTrailers(Http::RequestTrailerMap& trailers) {
+  return delegateFilterAction(delegated_filter_, &StreamDecoderFilter::decodeTrailers,
+                              Http::FilterTrailersStatus::Continue, trailers);
+}
+
+Http::FilterMetadataStatus Filter::decodeMetadata(Http::MetadataMap& metadata_map) {
+  return delegateFilterAction(delegated_filter_, &StreamDecoderFilter::decodeMetadata,
+                              Http::FilterMetadataStatus::Continue, metadata_map);
+}
+
+void Filter::decodeComplete() {
+  if (delegated_filter_) {
+    delegated_filter_->decodeComplete();
+  }
+}
+
+Http::FilterHeadersStatus Filter::encode100ContinueHeaders(Http::ResponseHeaderMap& headers) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encode100ContinueHeaders,
+                              Http::FilterHeadersStatus::Continue, headers);
+}
+
+Http::FilterHeadersStatus Filter::encodeHeaders(Http::ResponseHeaderMap& headers, bool end_stream) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encodeHeaders,
+                              Http::FilterHeadersStatus::Continue, headers, end_stream);
+}
+Http::FilterDataStatus Filter::encodeData(Buffer::Instance& data, bool end_stream) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encodeData,
+                              Http::FilterDataStatus::Continue, data, end_stream);
+}
+Http::FilterTrailersStatus Filter::encodeTrailers(Http::ResponseTrailerMap& trailers) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encodeTrailers,
+                              Http::FilterTrailersStatus::Continue, trailers);
+}
+Http::FilterMetadataStatus Filter::encodeMetadata(Http::MetadataMap& metadata_map) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encodeMetadata,
+                              Http::FilterMetadataStatus::Continue, metadata_map);
+}
+void Filter::encodeComplete() {
+  if (delegated_filter_) {
+    delegated_filter_->encodeComplete();
+  }
+}
+void Filter::onMatchCallback(const Matcher::Action& action) {
+  const auto& composite_action = action.getTyped<CompositeAction>();
+
+  FactoryCallbacksWrapper wrapper(*this);
+  composite_action.createFilters(wrapper);
+
+  if (!wrapper.errors_.empty()) {
+    ENVOY_LOG(error, "failed to create delegated filter {}",
```

Comment:
This can spam on the data path, so I would make this debug or a rate limited WARN (if we have that now) and have a stat.